src/Microsoft.Health.Dicom.Core/Features/Update/UpdateInstanceService.cs

```diff
@@ -0,0 +1,202 @@
+        ILogger<UpdateInstanceService> logger)
+    {
+        _fileStore = EnsureArg.IsNotNull(fileStore, nameof(fileStore));
+        _metadataStore = EnsureArg.IsNotNull(metadataStore, nameof(metadataStore));
+        _logger = EnsureArg.IsNotNull(logger, nameof(logger));
+        _recyclableMemoryStreamManager = EnsureArg.IsNotNull(recyclableMemoryStreamManager, nameof(recyclableMemoryStreamManager));
+    }
+
+    /// <inheritdoc />
+    public async Task UpdateInstanceBlobAsync(InstanceFileIdentifier instanceFileIdentifier, DicomDataset datasetToUpdate, CancellationToken cancellationToken)
+    {
+        EnsureArg.IsNotNull(datasetToUpdate, nameof(datasetToUpdate));
+        EnsureArg.IsNotNull(instanceFileIdentifier, nameof(instanceFileIdentifier));
+        EnsureArg.IsTrue(instanceFileIdentifier.NewVersion.HasValue, nameof(instanceFileIdentifier.NewVersion.HasValue));
+
+        Task updateInstanceFileTask = UpdateInstanceFileAsync(instanceFileIdentifier, datasetToUpdate, cancellationToken);
+        Task updateInstanceMetadataTask = UpdateInstanceMetadataAsync(instanceFileIdentifier, datasetToUpdate, cancellationToken);
+        await Task.WhenAll(updateInstanceFileTask, updateInstanceMetadataTask);
+    }
+
+    /// <inheritdoc />
+    public async Task DeleteInstanceBlobAsync(long fileIdentifier, CancellationToken cancellationToken = default)
+    {
+        _logger.LogInformation("Begin deleting instance blob {FileIdentifier}.", fileIdentifier);
+
+        Task fileTask = _fileStore.DeleteFileIfExistsAsync(fileIdentifier, cancellationToken);
+        Task metadataTask = _metadataStore.DeleteInstanceMetadataIfExistsAsync(fileIdentifier, cancellationToken);
+
+        await Task.WhenAll(fileTask, metadataTask);
+
+        _logger.LogInformation("Deleting instance blob {FileIdentifier} completed successfully.", fileIdentifier);
+    }
+
+    private async Task UpdateInstanceFileAsync(InstanceFileIdentifier instanceFileIdentifier, DicomDataset datasetToUpdate, CancellationToken cancellationToken)
+    {
+        long originFileIdentifier = instanceFileIdentifier.Version;
+        long newFileIdentifier = instanceFileIdentifier.NewVersion.Value;
+
+        KeyValuePair<string, long> block = default;
+
+        // If the file is already updated, then we can copy the file and just update the first block
+        if (instanceFileIdentifier.OriginalVersion.HasValue)
+        {
+            _logger.LogInformation("Begin copying instance file {OrignalFileIdentifier} - {NewFileIdentifier}", originFileIdentifier, newFileIdentifier);
+            await _fileStore.CopyFileAsync(originFileIdentifier, newFileIdentifier, cancellationToken);
+        }
+        else
+        {
+            _logger.LogInformation("Begin downloading original file {OrignalFileIdentifier} - {NewFileIdentifier}", originFileIdentifier, newFileIdentifier);
+
+            using Stream stream = await _fileStore.GetFileAsync(originFileIdentifier, cancellationToken);
```

Comment:
This is bringing the entire file in-memory. You need a FileStream

Reply:
I thought it opens the connection and buffers 4MB chunks to the stream, since we are opening the file as large file on demand, we may not read the whole file right? And we also upload the the content to clone blob in 4 MB blobs so that means is the file entirely stored in memory? I will run some tests to understand

Reply:
Discussed offline. 
The blob api indeed downloads only in chunks so it is not downloading the entire file, but the DicomFile.OpenAsync(stream, ReadOption.LargeFileOnDemand) puts everything in memory.

If we use ReadOption.SkipLargeTags then we will not know when to stop to get the cursor before the large file. but if we use

DicomFile.OpenAsync(stream, ReadOption.LargeFileOnDemand, largeObjectSize: 1000) then fo-dicom buffers it while calling GetBytes and I am still able to get the offset

Overall the memory allocated was max 75MB for this application